text/0715-implicit-application-query-params.md

```diff
@@ -0,0 +1,117 @@
+---
+Stage: Accepted
+Start Date: 2020-01-30
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/715
+---
+
+<!---
+Directions for above:
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Arbitrary Query Params
+
+## Summary
+
+This RFC aims to add opt-in behavior that would allow developers to use *any*
+Query Param passed to `transitionTo`. This will hopefully improve the ergonomics
+of using query params in most default situations.
+
+## Motivation
+
+Query Params require configuration to use today, which is counter to our
+"convention over configuration" philosophy. This RFC proposes an opt-in change to
+allow for query params to _always_ be used in transitions and links, and no longer
+require the allow-list that controllers currently implement today.
+
+## Detailed design
+
+Arbitrary Query Params will need to be guarded by an optional feature, because:
+ - controllers are allow-list only, and the ability to allow all query params
+   by default conflicts with the feature set of a controller's allow-list.
+ - query param values will not be stored on any controller, whereas the
+   controller is the only place tracked-access to query params exists.
+
+
+### Example
+
+Query Params as they work today are allow-list only, and block all other Query Params.
+
+Example:
+```js
+export default class ArticlesController extends Controller {
+  queryParams = ['category'];
+}
+```
+
+With `queryParams` specified, only `category` will be allowed on the route for
+this controller.
+
+This RFC only proposes a change for this scenario:
+```js
+// Empty or default controller
+export default class ArticlesController extends Controller {
+}
+```
+
+where `<RouterService>#transitionTo({ queryParams: { category: 'anything' } })`
+and LinkTo will _not_ strip away the `category` Query Param.
+
+### Migration Path
+
+1. Update optional-features.json
+
+    ```
+    // config/optional-features.json
+    {
+      "arbitrary-query-params": true
+    }
+    ```
+
+2. Delete Query Param allow lists
+
+    ```diff
+     export default class ArticlesController extends Controller {
+    -  queryParams = ['category'];
```

Comment:
There is some maybe undefined behavior around _not_ having this specified atm.

For example, `this.router.currentRoute.queryParams` is still updated with transitions, but all QPs are stripped from the URL.

This RFC, to put it simply, makes all things equal:
 - transitionTo
 - matches URL
 - matches router service state
 - does not set properties on controllers
 
 That's it.

Reply:
Here is a demonstration of what happens today when the query params array is omitted: https://ember-twiddle.com/7e472191b3f5021433b8552158a4379e?openFiles=routes.articles%5C.js%2C&route=%2Farticles

(this demo mostly demonstrates how to manage your own sticky params with Ember 3.18, but there is weird router behavior demonstrated, too)